dot/node.go

```diff
@@ -35,15 +41,92 @@ var logger = log.NewFromGlobal(log.AddContext("pkg", "dot"))
 // Node is a container for all the components of a node.
 type Node struct {
 	Name          string
-	Services      *services.ServiceRegistry // registry of all node services
+	Services      services.ServiceRegistryIFace // registry of all node services
 	wg            sync.WaitGroup
 	started       chan struct{}
 	metricsServer *metrics.Server
 }
 
+//go:generate mockgen -source=node.go -destination=mock_node_builder_test.go -package=$GOPACKAGE
+
+type nodeBuilderIface interface {
+	nodeInitialised(string) error
+	initNode(config *Config) error
+	createStateService(config *Config) (*state.Service, error)
+	createNetworkService(cfg *Config, stateSrvc *state.Service, telemetryMailer telemetry.Client) (*network.Service,
+		error)
+	createRuntimeStorage(st *state.Service) (*runtime.NodeStorage, error)
+	loadRuntime(cfg *Config, ns *runtime.NodeStorage, stateSrvc *state.Service, ks *keystore.GlobalKeystore,
+		net *network.Service) error
+	createBlockVerifier(st *state.Service) (*babe.VerificationManager, error)
+	createDigestHandler(lvl log.Level, st *state.Service) (*digest.Handler, error)
+	createCoreService(cfg *Config, ks *keystore.GlobalKeystore, st *state.Service, net *network.Service,
+		dh *digest.Handler) (*core.Service, error)
+	createGRANDPAService(cfg *Config, st *state.Service, dh *digest.Handler, ks keystore.Keystore,
+		net *network.Service, telemetryMailer telemetry.Client) (*grandpa.Service, error)
+	newSyncService(cfg *Config, st *state.Service, fg dotsync.FinalityGadget, verifier *babe.VerificationManager,
+		cs *core.Service, net *network.Service, telemetryMailer telemetry.Client) (*dotsync.Service, error)
+	createBABEService(cfg *Config, st *state.Service, ks keystore.Keystore, cs *core.Service,
+		telemetryMailer telemetry.Client) (babe.ServiceIFace, error)
+	createSystemService(cfg *types.SystemInfo, stateSrvc *state.Service) (*system.Service, error)
+	createRPCService(params rpcServiceSettings) (*rpc.HTTPServer, error)
+}
+
+var _ nodeBuilderIface = (*nodeBuilder)(nil)
+
+type nodeBuilder struct{}
+
+// NodeInitialized returns true if, within the configured data directory for the
+// node, the state database has been created and the genesis data has been loaded
+func NodeInitialized(basepath string) bool {
+	nodeInstance := nodeBuilder{}
+	err := nodeInstance.nodeInitialised(basepath)
+	if err != nil {
+		logger.Errorf("failed to initialise node from base path %s: %s", basepath, err)
+		return false
+	}
+	return true
+}
+
+func (*nodeBuilder) nodeInitialised(basepath string) error {
+	// check if key registry exists
+	registry := path.Join(basepath, utils.DefaultDatabaseDir, "KEYREGISTRY")
+
+	_, err := os.Stat(registry)
+	if os.IsNotExist(err) {
+		return err
+	}
+
+	// initialise database using data directory
+	db, err := utils.SetupDatabase(basepath, false)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err = db.Close()
+		if err != nil {
+			logger.Errorf("failed to close database: %s", err)
```

Comment:
otherwise it overrides the returned error

```suggestion
		closeErr := db.Close()
		if closeErr != nil {
			logger.Errorf("failed to close database: %s", closeErr)
```

Reply:
Updated.